Add fullscreen option.

de-vri-es opened the pull request:
This PR adds an option to enter fullscreen mode, hiding all interface elements and showing nothing but the rendered view. It makes rviz much nicer to use for showing information in demo setups.

I added the option under the `view_menu_` in code, but it turns out it's actually called "Panels" in the GUI. That might make it the wrong place for this option. Alternatively the menu could be renamed to view, or a new view menu could be added.

F11 also works, and is the only way other than restarting rviz to leave fullscreen mode. Might make sense to also use Escape for that?

One point of note: I'm using `setHideButtonVisibility(bool)` which is a public member of `VisualizationFrame`. I couldn't find anything actually calling the member, but if it is exposed to plugins they could potentially interact badly with the fullscreen option. It seems unlikely though, so maybe it's better to make that member private?

de-vri-es commented:
Oh, I tested on a system runing xmonad without window decorations, so I'm not 100% that `showFullScreen()` removes the window decoration. I kind of assumed it does.

de-vri-es commented:
Updated to use signals/slot for notifying `PanelDockWidgets` of full screen changes to decouple them from the `VisualizationFrame` again.

wjwwood commented:
@ros-pull-request-builder retest this please

wjwwood commented:
@ros-pull-request-builder retest this please

VictorLamoine commented:
I tested this PR and it works perfectly.

Minor issue: the maximized window state is not retrieved when exiting full screen (the window is not maximized anymore).

Really useful for demos/screenshots.

:+1:

de-vri-es commented:
I'll see if there is a sensible way to retain maximized state. Thanks for testing!

de-vri-es commented:
By the way, escape does work to exit full screen mode.

wjwwood commented:
Did you ever find a way to preserve the maximized state?

VictorLamoine commented on src/rviz/visualization_frame.h:
```diff
@@ -350,6 +358,9 @@ protected Q_SLOTS:
   QString error_message_; ///< Error message (if any) from most recent saveDisplayConfig() call.
+
+  /// Indicates if the toolbar should be visible outside of fullscreen mode.
+  bool toolbar_visibile_;
```
Typo, `visible`

de-vri-es replied:
Thanks btw, fixed.

VictorLamoine commented:
1 - Retrieve if the window was [maximized]( http://doc.qt.io/qt-5.7/qwidget.html#maximized-prop) before going full screen
2 - When returning from full screen, call [`showMaximized`](http://doc.qt.io/qt-4.8/qwidget.html#showMaximized) instread of `showNormal` if the window was maximized before.

The problem is that the `maximized` property does not work with X11. I tested and it is always zero.

de-vri-es commented:
There's also `QtWidget::saveGeometry()` and `QtWidget::restoreGeometry()`. Not sure how reliable that will be, but it's worth a try. Note that I'm currently on holiday, so earliest I can try it is somewhere next week.

VictorLamoine commented:
Fixes #743

de-vri-es commented:
Digging through the Qt documentation once more, I found this method of setting/disabling fullscreen:
```
if (full_screen)
  setWindowState(windowState() | Qt::WindowFullScreen);
else
  setWindowState(windowState() & ~Qt::WindowFullScreen);
show();
```

This preserves all other flags such as minimized (not applicable here I think) and maximized state. I tested with a `xfwm4` as window manager, and it worked as expected. Window state is preserved between entering and leaving fullscreen.

Updated and rebased the PR.

But note that I don't think it fixes #743. All widgets except for the 3D display are hidden in full-screen mode. That probably includes camera displays.

wjwwood commented:
Cool, I'll have another pass at it asap.

de-vri-es commented:
No rush, I haven't been rushing either, but have you had the time already?

wjwwood commented:
I'll come around to it before the Lunar release of rviz.

de-vri-es commented:
Heh, shooting for the moon. Anyway, thanks :)

VictorLamoine commented:
I tested again and it still works great.

The issue with the full-screen mode not being kept is still here but does not happen the first time. I think this really is minor and should not prevent the merging of this PR.

Here is a nice video:
https://drive.google.com/open?id=0B4_0QfSy1tFqa29fMHZCTWwxbHc

de-vri-es commented:
> The issue with the full-screen mode not being kept is still here but does not happen the first time. I think this really is minor and should not prevent the merging of this PR.

That is weird. I tested with `xfwm` and the previous version did mess up maximized state after leaving full screen, but the current did not.

Though I agree its something that shouldn't be a big problem for merging. Since it's a problem with Qt / X11 I actually think it's better to ignore it. The proper place to fix it is there. Any workaround we introduce may end up breaking if Qt fixes the issue on their end, and we end up with legacy code which nobody really remembers the purpose of.

wjwwood commented:
Finally got around to testing this. Looks good, thanks!

wxmerkt commented:
Awesome, thanks! Will this be backported to indigo and released into the debians?

wjwwood commented:
@wxmerkt I was just going to put it into Kinetic and Lunar, but since you asked nicely (and it seems low risk) I'll back port it to Indigo too.

de-vri-es commented:
hmm, maybe I shouldn't have removed the branch? Yay, github has a restore branch button.

Thanks for merging.

wjwwood commented:
@de-vri-es that's ok, I'll manually handle cherry-picking for forward/back porting.

wxmerkt commented:
Thank you @wjwwood :)

wjwwood commented:
Backported in https://github.com/ros-visualization/rviz/pull/1110
